src/Storages/StorageMerge.cpp

```diff
@@ -538,6 +538,14 @@ void StorageMerge::convertingSourceStream(
     }
 }
 
+IStorage::ColumnSizeByName StorageMerge::getColumnSizes() const
+{
+
+    auto first_materialize_mysql = getFirstTable([](const StoragePtr & table) { return table && table->getName() == "MaterializeMySQL"; });
```

Comment:
Would it be better to merge multiple columns_size  in the same way as the query?

Reply:
Probably yes, but all storages except *MergeTree return empty column sizes list and `TreeRewriter` [use](https://github.com/ClickHouse/ClickHouse/blob/12f66fa82c07fdde4a2410d6aadd0835445b1b87/src/Interpreters/TreeRewriter.cpp#L478) data type size to choose the smallest column as a fallback, so it's not clear how to merge them when we have no information about `compressed_size` for some tables.

Actually, we can choose any physical column to get number of rows in block if no physical columns are queried, but we are trying to choose the smallest one to read less data and get better performance. It is the root case of the issue with `count()`: `_sign` column is usually the smallest one, so it's implicitly added to the list of columns to read. When `_sign` and `_version` are removed from the column sizes list, `TreeRewriter` will choose the smallest one from the rest columns even if it's a String column. However, if not the best one are chosen, it affects only performance, and I think that it does not matter for MaterializeMySQL because of additional `FINAL` and `WHERE`.